stdlib/public/Observation/Sources/Observation/Observed.swift

```diff
@@ -0,0 +1,221 @@
+    var state: _ManagedCriticalState<State>?
+    let emit: @isolated(any) @Sendable () throws(Failure) -> Element?
+    
+    // this is the primary implementation of the tracking
+    // it is bound to be called on the specified isolation of the construction
+    fileprivate static func trackEmission(isolation trackingIsolation: isolated (any Actor)?, state: _ManagedCriticalState<State>, emit: @escaping @isolated(any) @Sendable () throws(Failure) -> Element?) throws(Failure) -> Element? {
+      // this ferries in an intermediate form with Result to skip over `withObservationTracking` not handling errors being thrown
+      // particularly this case is that the error is also an iteration state transition data point (it terminates the sequence)
+      // so we need to hold that to get a chance to catch and clean-up
+      let result = withObservationTracking {
+        Result(catching: emit)
+      } onChange: { [state] in
+        // resume all cases where the awaiting continuations are awaiting a willSet
+        State.emitWillChange(state)
+      }
+      return try result.get()
+    }
+    
+    fileprivate mutating func terminate(throwing failure: Failure? = nil, id: Int) throws(Failure) -> Element? {
+      // this is purely defensive to any leaking out of iteration generation ids
+      state?.withCriticalRegion { state in
+        state.continuations.removeValue(forKey: id)
+      }?.resume()
+      // flag the sequence as terminal by nil'ing out the state
+      state = nil
+      if let failure {
+        throw failure
+      } else {
+        return nil
+      }
+    }
+    
+    fileprivate mutating func trackEmission(isolation iterationIsolation: isolated (any Actor)?, state: _ManagedCriticalState<State>, id: Int) async throws(Failure) -> Element? {
+      guard !Task.isCancelled else {
+        // the task was cancelled while awaiting a willChange so ensure a proper termination
+        return try terminate(id: id)
+      }
+      // start by directly tracking the emission via a withObservation tracking on the isolation specified fro mthe init
+      guard let element = try await Iterator.trackEmission(isolation: emit.isolation, state: state, emit: emit) else {
+        // the user returned a nil from the closure so terminate the sequence
+        return try terminate(id: id)
+      }
+      return element
+    }
+    
+    public mutating func next(isolation iterationIsolation: isolated (any Actor)? = #isolation) async throws(Failure) -> Element? {
+      // early exit if the sequence is terminal already
+      guard let state else { return nil }
+      // set up an id for this generation
+      let id = State.generation(state)
+      do {
+        // there are two versions;
+        // either the tracking has never yet started at all and we need to prime the pump
+        // or the tracking has already started and we are going to await a change
+        if State.startTracking(state) {
```

Comment:
maybe not of interest for an initial implementation, but it occurred to me that we could perhaps reduce the number of lock acquisitions by merging the tracking logic and id generation. we could maybe even even get away with dropping the separate storage field for the tracking flag too if we reserved a sentinel id value to mean 'not yet tracking'.

Reply:
no longer relevant with the latest changes